server/src/main/java/org/opensearch/index/recovery/RemoteStoreRestoreService.java

```diff
@@ -150,4 +188,80 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
+        boolean restoreAllShards,
+        String restoreClusterUUID
+    ) {
+        String errorMsg = "cannot restore index [%s] because an open index with same name/uuid already exists in the cluster.";
+
+        List<String> graveyardIndexNames = new ArrayList<>();
+        List<String> graveyardIndexUUID = new ArrayList<>();
+        List<String> liveClusterIndexUUIDs = currentState.metadata()
+            .indices()
+            .values()
+            .stream()
+            .map(IndexMetadata::getIndexUUID)
+            .collect(Collectors.toList());
+
+        currentState.metadata().indexGraveyard().getTombstones().forEach(tombstone -> {
+            graveyardIndexNames.add(tombstone.getIndex().getName());
+            graveyardIndexUUID.add(tombstone.getIndex().getUUID());
+        });
+        for (Map.Entry<String, Tuple<Boolean, IndexMetadata>> indexMetadataEntry : indexMetadataMap.entrySet()) {
+            String indexName = indexMetadataEntry.getKey();
+            IndexMetadata indexMetadata = indexMetadataEntry.getValue().v2();
+            String indexUUID = indexMetadata.getIndexUUID();
+            boolean fromRemoteStore = indexMetadataEntry.getValue().v1();
+            if (indexMetadata.getSettings().getAsBoolean(SETTING_REMOTE_STORE_ENABLED, false)) {
+                if (fromRemoteStore) {
+                    // Since updates to graveyard are synced to remote we should neven land in a situation where remote contain index
+                    // metadata for graveyard index.
+                    assert graveyardIndexNames.contains(indexName) == false : String.format(
+                        Locale.ROOT,
+                        "Index name [%s] exists in graveyard!",
+                        indexName
+                    );
+                    assert graveyardIndexUUID.contains(indexUUID) == false : String.format(
+                        Locale.ROOT,
+                        "Index UUID [%s] exists in graveyard!",
+                        indexUUID
+                    );
+
+                    // Restore with current cluster UUID will fail as same indices would be present in the cluster which we are trying to
+                    // restore
+                    if (currentState.metadata().clusterUUID().equals(restoreClusterUUID)) {
+                        String finalErrorMsg = "clusterUUID to restore from should be different from current cluster UUID";
+                        logger.info(finalErrorMsg);
+                        throw new IllegalArgumentException("clusterUUID to restore from should be different from current cluster UUID");
+                    }
+
+                    // Any indices being restored from remote cluster state should already be part of the cluster as this causes conflict
+                    boolean sameNameIndexExists = currentState.metadata().hasIndex(indexName);
+                    boolean sameUUIDIndexExists = liveClusterIndexUUIDs.contains(indexUUID);
```

Comment:
dont need a explicit check on UUID, checking on name is sufficient.

Reply:
uuid would be needed as remote indices use uuid in remote path